oozie/src/main/java/org/apache/falcon/workflow/engine/OozieWorkflowEngine.java

```diff
@@ -974,30 +965,61 @@ private void performAction(String cluster, JobAction action, CoordinatorAction c
         }
     }
 
-    private void reRunCoordAction(String cluster, CoordinatorAction coordinatorAction) throws FalconException {
+    public CoordinatorAction.Status reRunCoordAction(String cluster, CoordinatorAction coordinatorAction,
+                                                      Properties props, boolean isForced) throws FalconException {
         try {
             OozieClient client = OozieClientFactory.get(cluster);
+            if (props == null) {
+                props = new Properties();
+            }
+            Properties jobprops;
+            // Get conf when workflow is launched.
+            if (coordinatorAction.getExternalId() != null) {
+                WorkflowJob jobInfo = client.getJobInfo(coordinatorAction.getExternalId());
+
+                jobprops = OozieUtils.toProperties(jobInfo.getConf());
+                jobprops.putAll(props);
+                // In case if both props exists, throw an exception.
+                // This case will occur when user runs workflow with skip-nodes property and
+                // try to do force rerun or rerun with fail-nodes property.
+                if (props.containsKey(OozieClient.RERUN_FAIL_NODES)
+                        && props.containsKey(OozieClient.RERUN_SKIP_NODES)) {
+                    String msg = "Both " + OozieClient.RERUN_SKIP_NODES + " and " + OozieClient.RERUN_FAIL_NODES
+                            + " are present in workflow params for " + coordinatorAction.getExternalId();
+                    LOG.error(msg);
+                    throw new FalconException(msg);
+                }
+
+                //if user has set any of these oozie rerun properties then force rerun flag is ignored
+                if (props.containsKey(OozieClient.RERUN_FAIL_NODES)) {
```

Comment:
This condition might have problem, if user rerun with this property, he cannot do force rerun again. Force flag will be always set false.